src/suites/cts/robust_access.spec.ts

```diff
@@ -0,0 +1,364 @@
+      typedData.fill(42);
+      for (let i = 0; i < size / 4; i++) {
+        typedData[zeroStart / 4 + i] = 0;
+      }
+    },
+  },
+};
+
+// The definition of aggregate types.
+interface Type {
+  // String to declare a variable named "data" of that type.
+  declaration: string;
+  // The array length, which also defines the bounds for this type.
+  length: number;
+  // The footprints in the buffer of this type, in baseTypes
+  std140Length: number;
+  std430Length: number;
+  // String to produce this type filled with zeroes.
+  zero: string;
+  baseType: BaseType;
+}
+
+interface TypeDictionary {
+  [key: string]: Type;
+}
+
+const typeParams: TypeDictionary = (() => {
+  const types: TypeDictionary = {};
+  for (const baseTypeName of Object.keys(baseTypes)) {
+    const baseType = baseTypes[baseTypeName];
+
+    // Arrays
+    types[`${baseTypeName}_sizedArray`] = {
+      declaration: `${baseTypeName} data[3]`,
+      length: 3,
+      std140Length: 2 * 4 + 1,
+      std430Length: 3,
+      zero: baseType.glslZero,
+      baseType,
+    };
+    types[`${baseTypeName}_unsizedArray`] = {
+      declaration: `${baseTypeName} data[]`,
+      length: 3,
+      std140Length: 0, // Unused
+      std430Length: 3,
+      zero: baseType.glslZero,
+      baseType,
+    };
+
+    // Vectors
+    for (let dimension = 2; dimension <= 4; dimension++) {
+      types[`${baseTypeName}_vector${dimension}`] = {
+        declaration: `${baseType.glslPrefix}vec${dimension} data`,
+        length: dimension,
+        std140Length: dimension,
+        std430Length: dimension,
+        zero: baseType.glslZero,
+        baseType,
+      };
+    }
+  }
+
+  // Matrices, there are only float matrics in GLSL.
+  for (const transposed of [true, false]) {
+    for (let firstDim = 2; firstDim <= 4; firstDim++) {
+      for (let secondDim = 2; secondDim <= 4; secondDim++) {
+        const majorDim = transposed ? secondDim : firstDim;
+        const minorDim = transposed ? firstDim : secondDim;
+
+        const std140SizePerMinorDim = 4;
+        const std430SizePerMinorDim = minorDim === 3 ? 4 : minorDim;
+
+        let typeName = `mat${firstDim}`;
+        if (firstDim !== secondDim) {
+          typeName += `x${secondDim}`;
+        }
+
+        types[(transposed ? 'transposed_' : '') + typeName] = {
+          declaration: (transposed ? 'layout(row_major) ' : '') + `${typeName} data`,
+          length: firstDim * secondDim,
+          std140Length: std140SizePerMinorDim * (majorDim - 1) + minorDim,
+          std430Length: std430SizePerMinorDim * (majorDim - 1) + minorDim,
+          zero: `vec${secondDim}(0.0f)`,
+          baseType: baseTypes['float'],
+        };
+      }
+    }
+  }
+
+  return types;
+})();
+
+g.test('bufferMemory', async t => {
+  await t.initGLSL();
+
+  const type = typeParams[t.params.type];
+  const baseType = type.baseType;
+
+  const indicesToTest = [
+    // Check exact bounds
+    `-1 * one`,
```

Comment:
nit: could move the `* one` out from here to the top of the indicesToTest loop (`const index = '(' + indexValue + ') * one';`)

Also, should we test with constants as well (i.e. omitting the `* one`)?

Reply:
Did the test with constants but GLSLang yells about the OOB access, so I think we can't do it for now.

Reply:
Ack